File: HyQMOM.jl/test/test_realizability.jl
const TOL = 1e-10

@testset "Realizability" begin
    
    @testset "realizability_S2 basic" begin
        S110, S101, S011 = 0.5, 0.3, 0.4
        
        S110r, S101r, S011r, S2r = realizability(:S2, S110, S101, S011)
        
        @test abs(S110r) <= 1.0 + TOL
        @test abs(S101r) <= 1.0 + TOL
        @test abs(S011r) <= 1.0 + TOL
        @test S2r >= 0.0
    end
    
    @testset "realizability_S111 basic" begin
        S110, S101, S011 = 0.5, 0.3, 0.4
        S210, S201, S120, S021, S102, S012 = 0.1, 0.2, 0.15, 0.25, 0.18, 0.22
        S111 = 0.3
        
        S111r = realizability(:S111, S110, S101, S011, S210, S201, S120, S021, S102, S012, S111)
        
        @test isfinite(S111r)
    end
    
    @testset "realizability_S210 basic" begin
        S110, S101, S011 = 0.5, 0.3, 0.4
        S300 = 0.1
        S210, S201 = 0.1, 0.2
        S400 = 3.0
        H200 = max(eps(), S400 - S300^2 - 1)
        beta = 1.0
        
        S210r, S201r = realizability(:S210, S110, S101, S011, S300, S210, S201, H200, beta)
        
        @test isfinite(S210r)
        @test isfinite(S201r)
    end
    
    @testset "realizability_S220 basic" begin
        S110 = 0.5
        S220 = 1.0
        A220 = 1.5
        
        S220r = realizability(:S220, S110, S220, A220)
        
        @test isfinite(S220r)
        @test abs(S220r) <= A220 + TOL
    end
    
    @testset "realizable_2D basic" begin
        S300, S400 = 0.1, 3.0
        S110, S210, S310 = 0.5, 0.1, 0.2
        S120, S220 = 0.15, 1.0
        S030, S130, S040 = 0.2, 0.25, 3.0
        
        S210r, S120r, S310r, S220r, S130r = realizability(Symbol("2D"), 
            S300, S400, S110, S210, S310, S120, S220, S030, S130, S040)
        
        @test all(isfinite.([S210r, S120r, S310r, S220r, S130r]))
    end
    
    @testset "Gaussian moments remain realizable" begin
        rho = 1.0
        u, v, w = 0.0, 0.0, 0.0
        T = 1.0
        
        M = InitializeM4_35(rho, u, v, w, T, 0.0, 0.0, T, 0.0, T)
        C4, S4 = M2CS4_35(M)
        
        # Extract 2nd-order moments
        S110, S101, S011 = S4[7], S4[17], S4[26]
        
        # Apply realizability
        S110r, S101r, S011r, S2r = realizability(:S2, S110, S101, S011)
        
        # Should not change for Gaussian
        @test S110r ~= S110 atol=TOL
        @test S101r ~= S101 atol=TOL
        @test S011r ~= S011 atol=TOL
    end
    
    @testset "Realizability preserves finiteness" begin
        # Test with potentially problematic values
        S110, S101, S011 = 0.99, 0.98, 0.97
        
        S110r, S101r, S011r, S2r = realizability(:S2, S110, S101, S011)
        
        @test all(isfinite.([S110r, S101r, S011r, S2r]))
        @test abs(S110r) <= 1.0 + TOL
        @test abs(S101r) <= 1.0 + TOL
        @test abs(S011r) <= 1.0 + TOL
    end
end

File: 3D_MATLAB/tests/test_realizability.m
function tests = test_realizability
tests = functiontests(localfunctions);
end

function setupOnce(testCase)
    script_dir = fileparts(fileparts(mfilename('fullpath')));
    addpath(fullfile(script_dir, 'src'));
    addpath(fullfile(script_dir, 'src', 'autogen'));
    addpath(fullfile(script_dir, 'tests', 'utils'));
    
    testCase.TestData.tol = 1e-4;
end

function test_realizability_S2_valid(testCase)
    S110 = 0.5;
    S101 = 0.3;
    S011 = 0.4;
    
    [S110r, S101r, S011r, S2r] = realizability('S2', S110, S101, S011);
    
    verifyGreaterThanOrEqual(testCase, S2r, -testCase.TestData.tol, 'S2 should be non-negative');
    verifyEqual(testCase, S110r, S110, 'AbsTol', testCase.TestData.tol, 'S110 should be unchanged');
    
    fprintf('PASS: realizability S2 valid test\n');
end

function test_realizability_S2_boundary(testCase)
    S110 = 0.99;
    S101 = 0.99;
    S011 = 0.99;
    
    [S110r, S101r, S011r, S2r] = realizability('S2', S110, S101, S011);
    
    verifyGreaterThanOrEqual(testCase, S2r, -testCase.TestData.tol, 'S2 should be non-negative');
    verifyLessThanOrEqual(testCase, abs(S110r), 1.0 + testCase.TestData.tol, 'S110 in bounds');
    verifyLessThanOrEqual(testCase, abs(S101r), 1.0 + testCase.TestData.tol, 'S101 in bounds');
    verifyLessThanOrEqual(testCase, abs(S011r), 1.0 + testCase.TestData.tol, 'S011 in bounds');
    
    fprintf('PASS: realizability S2 boundary test\n');
end

function test_realizability_2D_valid(testCase)
    S30 = 0.5; S40 = 3.5; S11 = 0.3;
    S21 = 0.2; S31 = 0.1; S12 = 0.15;
    S22 = 1.2; S03 = 0.4; S13 = 0.1; S04 = 3.2;
    
    [S21r, S12r, S31r, S22r, S13r] = realizability('2D', S30, S40, S11, S21, S31, S12, S22, S03, S13, S04);
    
    verifyTrue(testCase, isfinite(S21r), 'S21r should be finite');
    verifyTrue(testCase, isfinite(S22r), 'S22r should be finite');
    
    fprintf('PASS: realizability 2D valid test\n');
end

function test_realizability_S111(testCase)
    S110 = 0.5; S101 = 0.3; S011 = 0.4;
    S210 = 0.2; S201 = 0.15; S120 = 0.18;
    S021 = 0.12; S102 = 0.14; S012 = 0.16;
    S111 = 0.1;
    
    S111r = realizability('S111', S110, S101, S011, S210, S201, S120, S021, S102, S012, S111);
    
    verifyTrue(testCase, isfinite(S111r), 'S111r should be finite');
    
    fprintf('PASS: realizability S111 test\n');
end

function test_edge_corner_correction_integration(testCase)
    S110r = 0.99; S101r = 0.3; S011r = 0.4;
    R110 = -0.01; R101 = 1 - S101r^2; R011 = 1 - S011r^2;
    
    [S300r1, S030r1, S400r1, S040r1] = deal(0.5, 0.6, 3.0, 3.2);
    [S300r2, S003r2, S400r2, S004r2] = deal(0.5, 0.5, 3.0, 3.1);
    [S030r3, S003r3, S040r3, S004r3] = deal(0.6, 0.5, 3.2, 3.1);
    [S300r, S030r, S003r, S400r, S040r, S004r] = deal(0.5, 0.6, 0.5, 3.0, 3.2, 3.1);
    [S210r, S201r, S120r, S111r, S102r, S021r, S012r] = deal(0.1, 0.1, 0.1, 0.05, 0.1, 0.1, 0.1);
    [S310r, S301r, S220r, S211r, S202r, S130r, S121r, S112r, S103r] = deal(0.05, 0.05, 1.2, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05);
    [S031r, S022r, S013r] = deal(0.05, 0.05, 0.05);
    
    [S110, S101, S011, S300, S030, S003, S400, S040, S004, ...
     S210, S201, S120, S111, S102, S021, S012, ...
     S310, S301, S220, S211, S202, S130, S121, S112, S103, S031, S022, S013] = ...
        edge_corner_correction(R110, R101, R011, S110r, S101r, S011r, ...
                               S300r1, S030r1, S400r1, S040r1, ...
                               S300r2, S003r2, S400r2, S004r2, ...
                               S030r3, S003r3, S040r3, S004r3, ...
                               S300r, S030r, S003r, S400r, S040r, S004r, ...
                               S210r, S201r, S120r, S111r, S102r, S021r, S012r, ...
                               S310r, S301r, S220r, S211r, S202r, S130r, S121r, S112r, S103r, ...
                               S031r, S022r, S013r);
    
    verifyEqual(testCase, abs(S110), 1.0, 'AbsTol', 0.01, 'S110 should be at edge');
    verifyTrue(testCase, isfinite(S300), 'S300 should be finite');
    
    S2 = 1 + 2*S110*S101*S011 - (S110^2 + S101^2 + S011^2);
    verifyGreaterThanOrEqual(testCase, S2, -testCase.TestData.tol, 'S2 realizability');
    
    fprintf('PASS: edge_corner_correction integration test\n');
end

function test_realizability_preserves_mass(testCase)
    rho = 1.5;
    u = 0.1; v = 0.2; w = 0.3;
    T = 1.0;
    
    M = InitializeM4_35(rho, u, v, w, T, 0, 0, T, 0, T);
    
    [~, ~, ~, M_real] = Flux_closure35_and_realizable_3D(M, 1, 0.5);
    
    verifyEqual(testCase, M_real(1), M(1), 'AbsTol', testCase.TestData.tol, ...
                'Realizability should preserve mass');
    
    fprintf('PASS: realizability preserves mass test\n');
end
